File: java/Decent.java

import java.util.Scanner;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Iterator;

public class Decent {
    private final String input;
    private int pos;
    private final Deque<String> callStack = new ArrayDeque<>();

    public Decent(String input) {
        this.input = input;
    }

    public boolean parse() {
        System.out.printf("%-20s %-20s %s%n", "Stack", "Input", "Action");
        callStack.push("E");
        print("Start");

        boolean accepted = E() && end();
        print(accepted ? "Accepted" : "Rejected");
        return accepted;
    }

    private boolean E() { print("E -> T E'"); return T() && EP(); }

    private boolean EP() {
        callStack.push("E'");
        if (!peek('+') && !peek('-')) { print("E' -> e"); callStack.pop(); return true; }
        char op = input.charAt(pos++); print("Match " + op);
        boolean ok = T() && EP();
        callStack.pop();
        return ok;
    }

    private boolean T() {
        callStack.push("T");
        print("T -> F T'");
        boolean ok = F() && TP();
        callStack.pop();
        return ok;
    }

    private boolean TP() {
        callStack.push("T'");
        if (!peek('*') && !peek('/')) { print("T' -> e"); callStack.pop(); return true; }
        char op = input.charAt(pos++); print("Match " + op);
        boolean ok = F() && TP();
        callStack.pop();
        return ok;
    }

    private boolean F() {
        callStack.push("F");
        boolean ok;
        if (match('(')) {
            print("Match (");
            ok = E() && match(')');
            if (ok) print("Match )");
        } else if (readId()) {
            print("Match id");
            ok = true;
        } else if (readNum()) {
            print("Match num");
            ok = true;
        } else {
            print("Error in F");
            ok = false;
        }
        callStack.pop();
        return ok;
    }

    private boolean readId() {
        skip();
        if (pos >= input.length()) return false;

        char c = input.charAt(pos);
        if (!Character.isLetter(c) && c != '_') return false;

        pos++;
        while (pos < input.length()) {
            char ch = input.charAt(pos);
            if (!Character.isLetterOrDigit(ch) && ch != '_') break;
            pos++;
        }
        return true;
    }

    private boolean readNum() {
        skip();
        int start = pos;
        boolean hasDot = false;

        while (pos < input.length()) {
            char c = input.charAt(pos);
            if (Character.isDigit(c)) pos++;
            else if (c == '.' && !hasDot) {
                hasDot = true;
                pos++;
            } else break;
        }
        return pos > start;
    }

    private boolean peek(char c) { skip(); return pos < input.length() && input.charAt(pos) == c; }

    private boolean match(char c) {
        if (peek(c)) {
            pos++;
            return true;
        }
        return false;
    }

    private boolean end() { skip(); return pos == input.length(); }

    private void skip() {
        while (pos < input.length() && Character.isWhitespace(input.charAt(pos))) {
            pos++;
        }
    }

    private void print(String action) {
        StringBuilder sb = new StringBuilder("$");
        Iterator<String> it = callStack.descendingIterator();
        while (it.hasNext()) sb.append(it.next());
        String inputView = input.substring(Math.min(pos, input.length())).trim() + "$";
        System.out.printf("%-20s %-20s %s%n", sb.toString(), inputView, action);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter expression (e.g., a+b*(c-2)):");
        String expression = sc.nextLine();

        Decent parser = new Decent(expression);
        parser.parse();
    }
}

File: cpp/decent.cpp
#include <algorithm>
#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

string inputExpr;
size_t pos = 0;
vector<string> callStack;

void skipSpaces() {
    while (pos < inputExpr.size() && isspace(static_cast<unsigned char>(inputExpr[pos]))) {
        pos++;
    }
}

bool peek(char c) {
    skipSpaces();
    return pos < inputExpr.size() && inputExpr[pos] == c;
}

bool match(char c) {
    if (peek(c)) {
        pos++;
        return true;
    }
    return false;
}

bool readId() {
    skipSpaces();
    if (pos >= inputExpr.size()) return false;

    char c = inputExpr[pos];
    if (!isalpha(static_cast<unsigned char>(c)) && c != '_') return false;

    pos++;
    while (pos < inputExpr.size()) {
        char ch = inputExpr[pos];
        if (!isalnum(static_cast<unsigned char>(ch)) && ch != '_') break;
        pos++;
    }
    return true;
}

bool readNum() {
    skipSpaces();
    size_t start = pos;
    bool hasDot = false;

    while (pos < inputExpr.size()) {
        char c = inputExpr[pos];
        if (isdigit(static_cast<unsigned char>(c))) {
            pos++;
        } else if (c == '.' && !hasDot) {
            hasDot = true;
            pos++;
        } else {
            break;
        }
    }

    return pos > start;
}

void printRow(const string& action) {
    string stackView = "$";
    for (auto it = callStack.rbegin(); it != callStack.rend(); ++it) {
        stackView += *it;
    }

    string inputView = inputExpr.substr(min(pos, inputExpr.size()));
    inputView.erase(inputView.begin(), find_if(inputView.begin(), inputView.end(), [](unsigned char ch) {
        return !isspace(ch);
    }));
    inputView += "$";

    cout << left << setw(20) << stackView << " "
         << left << setw(20) << inputView << " "
         << action << "\n";
}

bool E();
bool EP();
bool T();
bool TP();
bool F();

bool E() {
    printRow("E -> T E'");
    return T() && EP();
}

bool EP() {
    callStack.push_back("E'");
    if (!peek('+') && !peek('-')) {
        printRow("E' -> e");
        callStack.pop_back();
        return true;
    }

    char op = inputExpr[pos++];
    printRow(string("Match ") + op);
    bool ok = T() && EP();
    callStack.pop_back();
    return ok;
}

bool T() {
    callStack.push_back("T");
    printRow("T -> F T'");
    bool ok = F() && TP();
    callStack.pop_back();
    return ok;
}

bool TP() {
    callStack.push_back("T'");
    if (!peek('*') && !peek('/')) {
        printRow("T' -> e");
        callStack.pop_back();
        return true;
    }

    char op = inputExpr[pos++];
    printRow(string("Match ") + op);
    bool ok = F() && TP();
    callStack.pop_back();
    return ok;
}

bool F() {
    callStack.push_back("F");
    bool ok;

    if (match('(')) {
        printRow("Match (");
        ok = E() && match(')');
        if (ok) printRow("Match )");
    } else if (readId()) {
        printRow("Match id");
        ok = true;
    } else if (readNum()) {
        printRow("Match num");
        ok = true;
    } else {
        printRow("Error in F");
        ok = false;
    }

    callStack.pop_back();
    return ok;
}

bool endInput() {
    skipSpaces();
    return pos == inputExpr.size();
}

int main() {
    cout << "Enter expression (e.g., a+b*(c-2)):\n";
    getline(cin, inputExpr);

    cout << left << setw(20) << "Stack" << " "
         << left << setw(20) << "Input" << " "
         << "Action\n";

    callStack.push_back("E");
    printRow("Start");

    bool accepted = E() && endInput();
    printRow(accepted ? "Accepted" : "Rejected");
    return 0;
}



